cmcd/main.go

```diff
@@ -234,72 +176,89 @@ func main() {
 			return
 		}
 	}
-	metadata, certParams, pcrs, err := loadMetadata(c.LocalPath)
+	metadata, err := loadMetadata(c.LocalPath)
 	if err != nil {
 		log.Errorf("Failed to load metadata: %v", err)
 		return
 	}
 
-	// Check if the TPM is provisioned. If provisioned, load the AK and TLS key.
-	// Otherwise perform credential activation with provisioning server and then load the keys
-	provisioningRequired, err := tpmdriver.IsTpmProvisioningRequired(c.akPath)
-	if err != nil {
-		log.Error("Failed to check if TPM is provisioned - ", err)
-		return
-	}
-
-	err = tpmdriver.OpenTpm()
-	if err != nil {
-		log.Error("Failed to open the TPM. Check if you have privileges to open /dev/tpm0 - ", err)
-		return
-	}
-	defer tpmdriver.CloseTpm()
-
-	if provisioningRequired {
-		paths := &tpmdriver.Paths{
-			Ak:          c.akPath,
-			TLSKey:      c.tlsKeyPath,
-			AkCert:      c.akCertPath,
-			TLSCert:     c.tlsCertPath,
-			DeviceSubCa: c.deviceSubCaPath,
-			Ca:          c.caPath,
+	verifyingCerts := make([]byte, 0)
+	var tpm *tpmdriver.Tpm
+	var snp *snpdriver.Snp
+	var sw *swdriver.Sw
+
+	measurements := make([]ar.Measurement, 0)
+	var signer ar.Signer
+
+	if c.SigningInterface == "TPM" || contains("TPM", c.MeasurementInterfaces) {
+		tpmConfig := &tpmdriver.Config{
+			Paths: tpmdriver.Paths{
+				Ak:          c.akPath,
+				TLSKey:      c.tlsKeyPath,
+				AkCert:      c.akCertPath,
+				TLSCert:     c.tlsCertPath,
+				DeviceSubCa: c.deviceSubCaPath,
+				Ca:          c.caPath,
+			},
+			ServerAddr: c.ServerAddr,
+			KeyConfig:  c.KeyConfig,
+			Metadata:   metadata,
+			UseIma:     c.UseIma,
+			ImaPcr:     c.ImaPcr,
 		}
-		err = tpmdriver.ProvisionTpm(c.ServerAddr+"activate-credential/", paths, certParams, c.KeyConfig)
+
+		tpm, err = tpmdriver.NewTpm(tpmConfig)
 		if err != nil {
-			log.Error("Failed to provision TPM - ", err)
-			os.Remove(c.akPath)
-			os.Remove(c.tlsKeyPath)
+			log.Errorf("Failed to create new TPM driver: %v", err)
 			return
 		}
-	} else {
-		err = tpmdriver.LoadTpmKeys(c.akPath, c.tlsKeyPath)
+		defer tpmdriver.CloseTpm()
+	}
+
+	if contains("TPM", c.MeasurementInterfaces) {
+		log.Info("Using TPM as Measurement Interface")
+		measurements = append(measurements, tpm)
+	}
+
+	if c.SigningInterface == "TPM" {
+		log.Info("Using TPM as Signing Interface")
+		verifyingCerts = append(verifyingCerts, tpm.GetCertChain().Ca...)
+		signer = tpm
+	}
+
+	if contains("SNP", c.MeasurementInterfaces) {
+		log.Info("Using SNP as Measurement Interface")
+		snp, err = snpdriver.NewSnpDriver()
 		if err != nil {
-			log.Error("Failed to load TPM keys - ", err)
+			log.Errorf("Failed to create new SNP driver: %v", err)
 			return
 		}
+
+		measurements = append(measurements, snp)
 	}
 
-	certs, err := loadCerts(c)
-	if err != nil {
-		log.Errorf("Failed to load certificates: %v", err)
-		return
+	if c.SigningInterface == "SNP" {
```

Comment:
this is not yet supported in the current PR, right? should we then remove lines 240-242 to not have an "empty if" lying around?

Reply:
done